base/staticint.jl

```diff
@@ -0,0 +1,357 @@
+
+"""
+    StaticInt(N::Int) -> StaticInt{N}()
+
+A statically sized `Int`. Use `StaticInt(N)` instead of `Val(N)` when you want it to behave like a number.
+"""
+struct StaticInt{N} <: Integer
+    StaticInt{N}() where {N} = new{N::Int}()
+    StaticInt(N::Int) = StaticInt{N}()
+    StaticInt(N::StaticInt) = N
+    StaticInt(::Val{N}) where {N} = StaticInt(N)
+    StaticInt(N) = StaticInt(Int(N))
+end
+
+_dynamic_int(@nospecialize(x::StaticInt))::Int = _dynamic_int(typeof(x))
+_dynamic_int(@nospecialize(x::Type{<:StaticInt}))::Int = x.parameters[1]
+
+const Zero = StaticInt{0}
+const One = StaticInt{1}
+const IntType = Union{Int,StaticInt}
+IntType(x::Integer) = Int(x)
+IntType(x::IntType) = x
+
+convert(::Type{T}, @nospecialize(N::StaticInt)) where {T<:Number} = convert(T, Int(N))
+Bool(x::StaticInt{0}) = false
+Bool(x::StaticInt{1}) = true
+
+BigInt(@nospecialize(x::StaticInt)) = BigInt(Int(x))
+Integer(x::StaticInt) = x
+(::Type{T})(@nospecialize(x::StaticInt)) where {T<:Integer} = T(_dynamic_int(x))
+(::Type{T})(x::Int) where {T<:StaticInt} = StaticInt(x)
+convert(::Type{StaticInt{N}}, ::StaticInt{N}) where {N} = StaticInt{N}()
+
+promote_rule(@nospecialize(T1::Type{<:StaticInt}), ::Type{T2}) where {T2<:Number} = promote_type(Int, T2)
+promote_rule(@nospecialize(T1::Type{<:StaticInt}), ::Type{T2}) where {T2<:AbstractIrrational} = promote_type(Int, T2)
+for (S, T) in [(:Complex, :Real), (:Rational, :Integer), (:(TwicePrecision), :Any)]
+    @eval function promote_rule(::Type{$S{T}}, @nospecialize(SI::Type{<:StaticInt})) where {T<:$T}
+        promote_type($S{T}, Int)
+    end
+end
+promote_rule(::Type{Union{Nothing,Missing}}, @nospecialize(T::Type{<:StaticInt})) = Union{Nothing,Missing,Int}
+promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Union{Missing,Nothing}} = promote_type(T1, Int)
+promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Nothing} = promote_type(T1, Int)
+promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Missing} = promote_type(T1, Int)
+for T in [:Bool, :Missing, :BigFloat, :BigInt, :Nothing, :Any]
+    # let S = :Any
+    @eval begin
+        promote_rule(@nospecialize(S::Type{<:StaticInt}), ::Type{$T}) = promote_type(Int, $T)
+        promote_rule(::Type{$T}, @nospecialize(S::Type{<:StaticInt})) = promote_type($T, Int)
+    end
+end
+promote_rule(@nospecialize(T1::Type{<:StaticInt}), @nospecialize(T2::Type{<:StaticInt})) = Int
+
+%(@nospecialize(n::StaticInt), ::Type{Integer}) = Int(n)
+
+eltype(@nospecialize(T::Type{<:StaticInt})) = Int
```

Comment:
Why make `eltype(StaticInt)` be `Int`? I assume it'd be identity by default? What would be wrong with that?

The reason I'm asking is that I have a package for (static) natural numbers here, and I'm considering having it subtype `Integer`: https://juliahub.com/ui/Packages/General/TypeDomainNaturalNumbers

BTW my package seems like a good alternative for the code here, because it's safer in two ways:

* It doesn't wrap other types, so there are no issues with needing to assume that the wrapped type is `Int`, like here
* negative numbers don't exist, which is safer for representing sizes

Reply:
Yeah, the invalidations are prohibitive. It seems any new `Integer` subtypes have to be compiled together with `Base`.